File: test/p2/main.js
function toProperCase(str) {
    // ensure that the SCREAMING_SNAKE_CASE is converted to snake_case
    if (str.match(/^[_A-Z0-9]+$/)) {
        str = str.toLowerCase();
    }

    // https://github.com/golang/lint/blob/5614ed5bae6fb75893070bdc0996a68765fdd275/lint.go#L771-L810
    const commonInitialisms = [
        "ACL", "API", "ASCII", "CPU", "CSS", "DNS", "EOF", "GUID", "HTML", "HTTP",
        "HTTPS", "ID", "IP", "JSON", "LHS", "QPS", "RAM", "RHS", "RPC", "SLA",
        "SMTP", "SQL", "SSH", "TCP", "TLS", "TTL", "UDP", "UI", "UID", "UUID",
        "URI", "URL", "UTF8", "VM", "XML", "XMPP", "XSRF", "XSS"
    ];

    return str.replace(/(^|[^a-zA-Z])([a-z]+)/g, function (unused, sep, frag) {
        if (commonInitialisms.indexOf(frag.toUpperCase()) >= 0)
            return sep + frag.toUpperCase();
        else
            return sep + frag[0].toUpperCase() + frag.substr(1).toLowerCase();
    }).replace(/([A-Z])([a-z]+)/g, function (unused, sep, frag) {
        if (commonInitialisms.indexOf(sep + frag.toUpperCase()) >= 0)
            return (sep + frag).toUpperCase();
        else
            return sep + frag;
    });
}


// // Example usage
// const inputString = "EXAMPLE_STRING";  // Replace this with your input string
// const properCaseString = toProperCase(inputString);
// console.log(properCaseString);

const inputs = [
    "$SCREAMING_SNAKE_CASE",
    "camelCase",
    "kebab-case",
    "Title Case",
    "lowercase",
    "HTML",
    "API",
    "HTTP",
    "CSS",
    "JSON",
    "URL"
];

inputs.forEach(input => {
    const result = toProperCase(input);
    console.log(`Input: ${input} => Output: ${result}`);
});

File: test/p2/main.go
package main

import (
	"fmt"
	"regexp"
	"strings"
)

func ToProperCase(str string) string {
	// Ensure that the SCREAMING_SNAKE_CASE is converted to snake_case
	if match, _ := regexp.MatchString("^[_A-Z0-9]+$", str); match {
		str = strings.ToLower(str)
	}

	// List of common initialisms
	commonInitialisms := map[string]bool{
		"ACL": true, "API": true, "ASCII": true, "CPU": true, "CSS": true, "DNS": true,
		"EOF": true, "GUID": true, "HTML": true, "HTTP": true, "HTTPS": true, "ID": true,
		"IP": true, "JSON": true, "LHS": true, "QPS": true, "RAM": true, "RHS": true,
		"RPC": true, "SLA": true, "SMTP": true, "SQL": true, "SSH": true, "TCP": true,
		"TLS": true, "TTL": true, "UDP": true, "UI": true, "UID": true, "UUID": true,
		"URI": true, "URL": true, "UTF8": true, "VM": true, "XML": true, "XMPP": true,
		"XSRF": true, "XSS": true,
	}

	// Convert the string to Proper Case
	re := regexp.MustCompile(`(^|[^a-zA-Z])([a-z]+)`)
	str = re.ReplaceAllStringFunc(str, func(match string) string {
		parts := re.FindStringSubmatch(match)
		sep, frag := parts[1], parts[2]

		if commonInitialisms[strings.ToUpper(frag)] {
			return sep + strings.ToUpper(frag)
		} else {
			return sep + strings.ToUpper(frag[0:1]) + strings.ToLower(frag[1:])
		}
	})

	re = regexp.MustCompile(`([A-Z])([a-z]+)`)
	str = re.ReplaceAllStringFunc(str, func(match string) string {
		parts := re.FindStringSubmatch(match)
		sep, frag := parts[1], parts[2]

		if commonInitialisms[sep+strings.ToUpper(frag)] {
			return (sep + frag)[0:]
		} else {
			return sep + frag
		}
	})

	return str
}

func main() {

	inputs := []string{
		"$SCREAMING_SNAKE_CASE",
		"camelCase",
		"kebab-case",
		"Title Case",
		"lowercase",
		"HTML",
		"API",
		"HTTP",
		"CSS",
		"JSON",
		"URL",
	}

	for _, input := range inputs {
		result := ToProperCase(input)
		fmt.Printf("Input: %s => Output: %s\n", input, result)
	}

}
